2018/python/rules_engine.py

```diff
@@ -305,32 +305,47 @@ def s305(self):
 		fail_df = self.lar_df[self.lar_df.duplicated(keep=False)==True] #pull frame of duplicates
 		self.results_wrapper(edit_name=edit_name, field_name=field, fail_df=fail_df)
 
-	def v608(self):
+	def v608_1(self):
 		"""A ULI with an invalid format was provided.
 		1) The required format for ULI is alphanumeric with at least 23 characters and up to 45 characters, 
 		and it cannot be left blank.
-		Exempt Filers:
-		If your institution is reporting a Loan/Application Number, the required format for Loan/Application Number 
-		is alphanumeric with at least 1 character and no more than 22 characters, and it cannot be left blank."""
-
-		edit_name = "v608"
+		"""
+		edit_name = "v608_1"
 		field = "ULI"
-		#if length not between 23 and 45 or if ULI is blank
-		#get subset of LAR dataframe that fails ULI conditions
-		#check if LEI present as first 20 digits of ULI
 		lei = self.lar_df.lei.iloc[0]
-		#get seperate dataframes for ULI and Loan ID 
-		uli_check_df = self.lar_df[(self.lar_df.uli.apply(lambda x: str(x)[:20]==lei))].copy()
-		loan_id_check_df = self.lar_df[(self.lar_df.uli.apply(lambda x: str(x)[:20]!=lei))].copy()
-		#filter each df for failures
-		uli_fail_df = uli_check_df[(uli_check_df.uli.map(lambda x: len(x)<23))|
-			(uli_check_df.uli.map(lambda x: len(x))>45)|(uli_check_df.uli=="")].copy()
-
-		loan_id_fail_df = loan_id_check_df[(loan_id_check_df.uli=="")|
-			(loan_id_check_df.uli.apply(lambda x: len(x)>22))].copy()
-		#recombine dfs
-		fail_df = pd.concat([uli_fail_df, loan_id_fail_df])
-		self.results_wrapper(edit_name=edit_name, field_name=field, fail_df=fail_df)
+		#Obtaining subset of LAR that would have ULIs rather than NULIs.
+		#Purchase loan ULIs would not necessarily contain the LEI in its first 20
+		#characters. 
+		if self.lar_df.action_taken.all() == '6':
```

Comment:
You still need to check ULIs on purchased loans. The check just needs to be LEI agnostic.